source/Detector/Scaler/AMDSScalerDetector.cpp

```diff
@@ -101,54 +147,110 @@ void AMDSScalerDetector::onDwellTimeControlValueChanged(double newValue)
 void AMDSScalerDetector::onScansInABufferControlValueChanged(double newValue)
 {
 	if (scansInABuffer_ != newValue) {
-		AMErrorMon::alert(this, 0, QString("Scaler %1 # scans in a buffer changed from %2 to %3.").arg(scalerName()).arg(scansInABuffer_).arg(newValue));
+		AMErrorMon::alert(this, AMDS_SCALER_DETECTOR_ALT_SCAN_IN_BUFFER_CHANGED, QString("Scaler %1 # scans in a buffer changed from %2 to %3.").arg(scalerName()).arg(scansInABuffer_).arg(newValue));
 
 		scansInABuffer_ = (int)newValue;
 	}
 }
 
-void AMDSScalerDetector::onContinuousScanControlValueChanged(double newValue)
+void AMDSScalerDetector::onTotalNumberOfScansControlValueChanged(double newValue)
+{
+	if (totalNumberOfScans_ != newValue) {
+		AMErrorMon::alert(this, AMDS_SCALER_DETECTOR_ALT_TOTAL_NUM_SCANS, QString("Scaler %1 total # scans changed from %2 to %3.").arg(scalerName()).arg(totalNumberOfScans_).arg(newValue));
+
+		totalNumberOfScans_ = (int)newValue;
+	}
+}
+
+void AMDSScalerDetector::onFetchScanBuffer()
 {
-	if ((newValue != AMDSScalerDetector::Continuous) && AMDSRunTimeSupport::debugAtLevel(1) ) {
-		AMErrorMon::alert(this, 0, QString("Scaler %1 switched to Normal scan mode.").arg(scalerName()));
+	if (!connected_)
+		return;
+
+	AMDSDataHolderList scanBufferDataHolderList;
+
+	int channelCount = enabledChannelCount();
+	QVector<int> countBuffer(channelCount * scansInABuffer_ + 1); // need one extra buffer for the # of channels
+	scanBufferControl_->values(channelCount, countBuffer.data());
+
+	// put the counts array to the FlatArrayList
+	for (int scanIndex=0; scanIndex < scansInABuffer_; scanIndex++) {
+		AMDSFlatArray channelCountArray(scalerConfiguration_->dataType(), channelCount);
+
+		int channelScanStartIndex = scanIndex * channelCount + 1;
+		channelCountArray.setValues(countBuffer.mid(channelScanStartIndex, channelCount));
```

Comment:
I think so. The data in the buffer is like:
```
#ofChannels v1 v2 v3 ... vn v1 v2 v3 ... vn v1 v2 v3 ... vn v1 v2 v3 ... vn
```
Therefore the start index of one scan is `scanIndex * channelCount + 1`, such as "1, n+1, 2n+1, 3n +1 ..."